portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,103 @@
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-
+  /**
+   * Retrieves comments from datastore, adds to array, and serves as JSON 
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    //Queries datastore for comments
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Comment").addSort("posted", SortDirection.ASCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    //Converts result from query into array of comment objects
+    ArrayList<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String name = (String) entity.getProperty("name");
+      String body = (String) entity.getProperty("body");
+      Date posted = (Date) entity.getProperty("posted");
+
+      Comment comment = new Comment(name, body ,posted);
+      comments.add(comment);
+    }
+
+    //Converts comments to JSON and sets response
+    String json = this.convertToJson(comments);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Stores comment in datastore
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String name = getParameter(request, "name", "Anonymous");
+    String body = getParameter(request, "body", "");
+
+    if (body.isEmpty()){
+      response.sendRedirect("/#comments");
+    }
+    else {
+      //Creates entity for comment
+      Entity commentEntity = new Entity("Comment");
+      commentEntity.setProperty("name", name);
+      commentEntity.setProperty("body", body);
+      commentEntity.setProperty("posted", new Date());
+
+      //Stores comment in datastore
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(commentEntity);
+
+      //Refreshes page
+      response.sendRedirect("/#comments");
+    }
+  }
+
+  /**
+   * Retrieves parameter, returning an alternative if param doesn't exist 
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null || value.isEmpty()) {
+      return defaultValue;
+    }
+    return sanitiseInput(value);
+  }
+
+  /**
+   * Sanitises input by replacing < and >
+   */
+  private String sanitiseInput(String input) {
+    String lessThan = input.replaceAll("<", "&lt;");
```

Comment:
I think GSON already does this (and more): https://stackoverflow.com/questions/23363843/gson-disablehtmlescaping-why-gson-html-escapes-by-default-in-the-first-place

Reply:
I've been experimenting with it a bit, and while it does escape the characters, the site still displays the entered html, though as far as I can tell it doesn't run <script> tags. I'll have to look into this more.

Reply:
Ack, good finding!